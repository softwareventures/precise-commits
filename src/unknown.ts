export function assertInstanceOf<TClass, TArgs extends unknown[]>(
    value: unknown,
    type: new (...args: TArgs) => TClass
): TClass {
    if (value instanceof type) {
        return value;
    } else {
        throw new TypeError(`Expected ${type.name}`);
    }
}
