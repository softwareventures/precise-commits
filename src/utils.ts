import * as execa from 'execa';

export interface LineChanges {
  start: number;
  noOfLines: number;
}

export interface LineChangeData {
  removals: LineChanges[];
  additions: LineChanges[];
}

export interface CharacterRange {
  rangeStart: number;
  rangeEnd: number;
}

export const NO_LINE_CHANGE_DATA_ERROR =
  'No line change data could be detected';

/**
 * Addition `start` number included in the range,
 * removal `start` is the line before
 */
export function extractLineChangeData(diffData: string) {
  const lineChanges = diffData.match(/@@.*@@/g);
  if (!lineChanges) {
    throw new Error(NO_LINE_CHANGE_DATA_ERROR);
  }
  const lineChangeData: {
    removals: LineChanges[];
    additions: LineChanges[];
  } = {
    removals: [],
    additions: [],
  };
  lineChanges.forEach(lineChange => {
    const d = lineChange.match(/(@@ )(-\d+,?\d*)( )(\+\d+,?\d*)( @@)/);
    if (!d) {
      throw new Error('The detected line change data could be not be parsed');
    }
    const [removalStartLine, noOfLinesRemoved = 1] = d[2].split(',');
    const [additionStartLine, noOfLinesAdded = 1] = d[4].split(',');
    if (noOfLinesRemoved > 0) {
      lineChangeData.removals.push({
        start: +removalStartLine.replace('-', ''),
        noOfLines: +noOfLinesRemoved,
      });
    }
    if (noOfLinesAdded > 0) {
      lineChangeData.additions.push({
        start: +additionStartLine.replace('+', ''),
        noOfLines: +noOfLinesAdded,
      });
    }
  });
  return lineChangeData;
}

function findCharacterIndexOfLine(
  fileContents: string,
  startCharIndex: number,
  lineCount: number,
): number {
  let charIndex = startCharIndex;
  let lineIndex = 0;
  while (lineIndex < lineCount && charIndex < fileContents.length) {
    const char = fileContents[charIndex];
    const nextChar = fileContents[charIndex + 1];
    if (char === '\n' || char === '\r') {
      ++lineIndex;
      if (char === '\r' && nextChar === '\n') {
        ++charIndex; // skip next character
      }
    }
    ++charIndex;
  }
  return charIndex;
}

export function calculateCharacterRangesFromLineChanges(
  lineChangeData: LineChangeData,
  fileContents: string,
): CharacterRange[] {
  let charIndex = 0;
  let lineIndex = 0;
  return lineChangeData.additions.map(added => {
    const rangeStart = findCharacterIndexOfLine(fileContents, charIndex, added.start - lineIndex - 1);
    const rangeEnd = findCharacterIndexOfLine(fileContents, rangeStart, added.noOfLines);
    charIndex = rangeEnd;
    lineIndex = added.start + added.noOfLines - 1;
    return { rangeStart, rangeEnd };
  });
}

export function runCommandSync(
  command: string,
  args: string[],
  workingDirectory = process.cwd(),
) {
  return execa.sync(command, args, { cwd: workingDirectory });
}

export function generateFilesWhitelistPredicate(
  filesWhitelist: string[] | null,
): (file: string) => boolean {
  if (!filesWhitelist) {
    return () => true;
  }
  return file => filesWhitelist.includes(file);
}
