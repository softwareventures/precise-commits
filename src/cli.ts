#!/usr/bin/env node

import {normalize} from "path";
import ora = require("ora");
import mri = require("mri");
import glob = require("glob");
import {hasProperty} from "unknown";
import {concatMap, isArray} from "@softwareventures/array";
import {pairwise} from "rxjs";
import {main} from ".";

const libraryName = "precise-commits";
const config = mri<unknown>(process.argv.slice(2));

/**
 * If the user provided one or more glob patterns to match against, ensure that there are
 * applicable files available
 */
let filesWhitelist: string[] | null = null;
if (hasProperty(config, "whitelist")) {
    if (isArray(config.whitelist)) {
        filesWhitelist = concatMap(config.whitelist, entry => glob.sync(String(entry)));
    } else {
        filesWhitelist = glob.sync(String(config.whitelist));
    }
    if ((filesWhitelist?.length ?? 0) === 0) {
        console.error(
            `Error: No files match the glob pattern(s) you provided for --whitelist -> "${String(
                config.whitelist
            )}"`
        );
        process.exit(1);
    }
    filesWhitelist = filesWhitelist.map(normalize);
}

/**
 * If the user specifies at least one SHA, perform some validation and
 * apply some defaults
 */
let base: string | null = null;
let head: string | null = null;
if (hasProperty(config, "base") || hasProperty(config, "head")) {
    if (!hasProperty(config, "base") || config.base == null) {
        console.error(
            `Error: When giving a value of --head, you must also give a value for --base`
        );
        process.exit(1);
    }

    if (hasProperty(config, "head") && config.head != null) {
        // FIXME: Tech debt
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        head = String(config.head);
    } else {
        /**
         * If the user only specified `--base`, set the value of `--head` to be "HEAD"
         */
        head = "HEAD";
    }

    // FIXME: Tech debt
    // eslint-disable-next-line @typescript-eslint/no-base-to-string
    base = String(config.base);
}

const checkOnly = hasProperty(config, "check-only") ? Boolean(config["check-only"]) : false;
const formatter = hasProperty(config, "formatter")
    ? // FIXME: Tech debt
      // eslint-disable-next-line @typescript-eslint/no-base-to-string
      String(config.formatter ?? "prettier")
    : "prettier";

const options = {
    checkOnly,
    filesWhitelist,
    base,
    head,
    formatter
};

const spinner = ora(`Running ${libraryName}`);

let shouldErrorOut = false;

main(process.cwd(), options)
    .pipe(pairwise())
    .subscribe({
        next: ([previous, current]) => {
            if (current.state === "Running") {
                if (previous.state === "Initializing") {
                    spinner.start(`Running ${libraryName}`);
                }

                if (current.files.length > 0) {
                    spinner.text = `${libraryName}: ${current.files.length} modified file(s) found`;
                }

                current.files.forEach((fileState, index) => {
                    const previousFileState =
                        previous.state === "Running" ? previous.files[index] : undefined;

                    if (previousFileState == null) {
                        spinner.info(`Processing file: ${fileState.filename}`);
                    }

                    if (previousFileState?.status !== fileState.status) {
                        if (fileState.status === "Updated") {
                            spinner.succeed(`Updated formatting in: ${fileState.filename}`);
                        } else if (fileState.status === "NotUpdated") {
                            spinner.succeed(
                                `No formatting changes required in: ${fileState.filename}`
                            );
                        } else if (fileState.status === "InvalidFormatting") {
                            // If --check-only is passed as a CLI argument, the script will error out.
                            if (options.checkOnly) {
                                shouldErrorOut = true;
                            }

                            spinner.fail(`Invalid formatting detected in: ${fileState.filename}`);
                        }
                    }
                });
            } else if (current.state === "Finished") {
                if (current.fileCount === 0) {
                    spinner.info("No matching modified files detected.");
                    spinner.info(
                        "If you feel that one or more files should be showing up here, be sure to first check what file extensions prettier supports, and whether or not you have included those files in a .prettierignore file"
                    );
                    spinner.stop();
                } else if (options.checkOnly) {
                    spinner.succeed("Checks complete ðŸŽ‰");
                } else {
                    spinner.succeed("Formatting complete ðŸŽ‰");
                }
                return process.exit(shouldErrorOut ? 1 : 0);
            }
        },
        error: (error: unknown) => {
            spinner.fail("An Error occurred");
            console.error("\n");
            console.error(error);
            console.error("\n");
            spinner.stop();
            return process.exit(1);
        }
    });
