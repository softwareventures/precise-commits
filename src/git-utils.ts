import {sync as findUpSync} from "find-up";
import {dirname} from "path";

import {runCommandSync} from "./utils";
import {hasProperty} from "unknown";

interface DiffIndexFile {
    diffFilterChar: string;
    filename: string;
}

export function resolveNearestGitDirectoryParent(workingDirectory: string) {
    const gitDirectoryPath = findUpSync(".git", {cwd: workingDirectory, type: "directory"});
    if (!gitDirectoryPath) {
        throw new Error("No .git directory found");
    }
    return dirname(gitDirectoryPath);
}

export function getDiffForFile(
    gitDirectoryParent: string,
    fullPath: string,
    base: string | null,
    head: string | null
): string {
    if (base && head) {
        return runCommandSync(
            "git",
            ["diff", "--unified=0", base, head, fullPath],
            gitDirectoryParent
        ).stdout;
    }
    return runCommandSync("git", ["diff", "--unified=0", "--cached", fullPath], gitDirectoryParent)
        .stdout;
}

/**
 * Output of `git diff-index --help`:
 *
 * --diff-filter=[(A|C|D|M|R|T|U|X|B)...[*]]
      Select only files that are Added (A), Copied (C), Deleted (D), Modified (M), Renamed (R), have
      their type (i.e. regular file, symlink, submodule, ...) changed (T), are Unmerged (U), are
      Unknown (X), or have had their pairing Broken (B). Any combination of the filter characters
      (including none) can be used. When * (All-or-none) is added to the combination, all paths are
      selected if there is any file that matches other criteria in the comparison; if there is no
      file that matches other criteria, nothing is selected.

      Also, these upper-case letters can be downcased to exclude. E.g.  --diff-filter=ad excludes
      added and deleted paths.
 *
 */
/**
 * const DIFF_INDEX_FILTER = 'ACDMRTUXB';
 * NOTE: We are only explicitly testing "Modified" and "Added" files for now...
 */
const DIFF_INDEX_FILTER = "AM";
const SPECIAL_EMPTY_TREE_COMMIT_HASH = "4b825dc642cb6eb9a060e54bf8d69288fbee4904";

export function getModifiedFilenames(
    gitDirectoryParent: string,
    base: string | null,
    head: string | null
): string[] {
    let diffIndexOutput: string;
    if (base && head) {
        /**
         * We are grabbing the files between the two given commit SHAs
         */
        diffIndexOutput = runCommandSync(
            "git",
            ["diff", "--name-status", `--diff-filter=${DIFF_INDEX_FILTER}`, base, head],
            gitDirectoryParent
        ).stdout;
    } else {
        /**
         * No commit SHAs given, we assume we are attempting to evaluate staged files,
         * and so need to determine if there is HEAD SHA available.
         */
        let head: string = "";
        try {
            head = runCommandSync(
                "git",
                ["rev-parse", "--verify", "HEAD"],
                gitDirectoryParent
            ).stdout.replace("\n", "");
        } catch (err) {
            /**
             * If there has never been a commit before, there will be no HEAD to compare
             * to. Use the special empty tree hash value instead:
             * https://stackoverflow.com/questions/9765453/is-gits-semi-secret-empty-tree-object-reliable-and-why-is-there-not-a-symbolic
             */
            if (
                hasProperty(err, "message") &&
                typeof err.message === "string" &&
                err.message.includes("fatal: Needed a single revision")
            ) {
                head = SPECIAL_EMPTY_TREE_COMMIT_HASH;
            } else {
                throw err;
            }
        }
        diffIndexOutput = runCommandSync(
            "git",
            ["diff-index", "--cached", "--name-status", `--diff-filter=${DIFF_INDEX_FILTER}`, head],
            gitDirectoryParent
        ).stdout;
    }
    const allFiles = parseDiffIndexOutput(diffIndexOutput);
    return allFiles.map(r => r.filename);
}

function parseDiffIndexOutput(stdout: string): DiffIndexFile[] {
    const lines = stdout.split("\n");
    return lines.filter(Boolean).map(line => {
        const parts = line.split("\t");
        return {
            filename: parts[1],
            diffFilterChar: parts[0]
        };
    });
}
