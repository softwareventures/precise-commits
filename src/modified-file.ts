import {readFileSync, writeFileSync} from "fs";
import {relative, sep, posix} from "path";
import {notNull} from "@softwareventures/nullable";
import execa = require("execa");
import type {CharacterRange} from "./utils";
import {
    calculateCharacterRangesFromLineChanges,
    extractLineChangeData
} from "./utils";
import {getDiffForFile, index, workingTree} from "./git-utils";
import type {PreciseFormatter} from "./precise-formatter";
import {assertInstanceOf} from "./unknown";

export interface ModifiedFileConfig {
    fullPath: string;
    gitDirectoryParent: string;
    base: string | null;
    head: string | typeof index | typeof workingTree;
    selectedFormatter: PreciseFormatter<any>;
}

export class ModifiedFile {
    private readonly fullPath: string;
    private readonly pathInGit: string;
    /**
     * An optional commit SHA pair which will be used to inform how the git
     * commands are run. E.g. `git diff`
     */
    private readonly base: string | null;
    private readonly head: string | typeof index | typeof workingTree;
    /**
     * The chosen formatter to be run on the modified file.
     */
    private readonly selectedFormatter: PreciseFormatter<any>;
    /**
     * The parent directory of the relevant .git directory that was resolved
     * for the modified file.
     */
    private readonly gitDirectoryParent: string;
    /**
     * The contents of the file in their current state on the user's file
     * system
     */
    private readonly fileContents: string;
    /**
     * The final file contents, after we've run the formatter
     */
    private formattedFileContents: string | null = null;
    /**
     * The resolved formatter config which applies to this file
     */
    private readonly formatterConfig: object | null;
    /**
     * The calculated character ranges which have been modified
     * within this file
     */
    private modifiedCharacterRanges: CharacterRange[] = [];

    constructor({fullPath, gitDirectoryParent, base, head, selectedFormatter}: ModifiedFileConfig) {
        this.fullPath = fullPath;
        this.pathInGit = relative(gitDirectoryParent, fullPath).split(sep).join(posix.sep);
        this.gitDirectoryParent = gitDirectoryParent;
        this.base = base;
        this.head = head;
        this.selectedFormatter = selectedFormatter;
        this.fileContents =
            head === workingTree
                ? readFileSync(this.fullPath, "utf8")
                : head === index
                ? execa.sync("git", ["show", `:0:${this.pathInGit}`], {
                      cwd: gitDirectoryParent,
                      stripFinalNewline: false
                  }).stdout
                : execa.sync("git", ["show", `${head}:${this.pathInGit}`], {
                      cwd: gitDirectoryParent,
                      stripFinalNewline: false
                  }).stdout;
        this.formatterConfig = this.selectedFormatter.resolveConfig(this.fullPath);
    }

    /**
     * Return true if the whole file has already been formatted appropriately based on
     * the resolved formatter config. We can use this as a check to skip unnecessary work.
     */
    isAlreadyFormatted(): boolean {
        return this.selectedFormatter.isAlreadyFormatted(this.fileContents, this.formatterConfig);
    }

    /**
     * Run the formatters check mode on the given ranges and return true if they are all
     * already formatted appropriately based on the resolved formatter config.
     */
    hasValidFormattingForCharacterRanges(): boolean {
        return this.selectedFormatter.checkFormattingOfRanges(
            this.fullPath,
            this.fileContents,
            this.formatterConfig,
            this.modifiedCharacterRanges
        );
    }

    /**
     * Run the formatter on the file contents and store the result
     */
    formatCharacterRangesWithinContents(): void {
        this.formattedFileContents = this.selectedFormatter.formatRanges(
            this.fullPath,
            this.fileContents,
            this.formatterConfig,
            this.modifiedCharacterRanges
        );
    }

    /**
     * Return true if the formatted file contents are different to
     * what was originally resolved from disk.
     */
    shouldContentsBeUpdatedOnDisk(): boolean {
        return this.fileContents !== this.formattedFileContents;
    }

    /**
     * Write the updated file contents back to disk.
     */
    updateFileOnDisk(): void {
        if (this.head === index) {
            const hash = execa.sync(
                "git",
                ["hash-object", "-w", "--path", this.fullPath, "--stdin"],
                {
                    cwd: this.gitDirectoryParent,
                    input: notNull(this.formattedFileContents)
                }
            ).stdout;
            const mode = execa
                .sync("git", ["ls-files", "--stage", "--", this.fullPath], {
                    cwd: this.gitDirectoryParent
                })
                .stdout.split(" ")?.[0];
            if (mode == null) {
                throw new Error("Can't find file in git index");
            }
            execa.sync(
                "git",
                [
                    "update-index",
                    "--add",
                    "--replace",
                    "--cacheinfo",
                    `100644,${hash},${this.pathInGit}`
                ],
                {cwd: this.gitDirectoryParent}
            );
        } else {
            writeFileSync(this.fullPath, notNull(this.formattedFileContents));
        }
    }

    /**
     * We handle errors locally within this method to allow for
     * more granular feedback within the main() function of the
     * library.
     */
    calculateModifiedCharacterRanges(): {err: Error | null} {
        try {
            /**
             * Extract line change data from the git diff results.
             */
            const diff = getDiffForFile(
                this.gitDirectoryParent,
                this.fullPath,
                this.base,
                this.head
            );
            const lineChangeData = extractLineChangeData(diff);
            /**
             * Convert the line change data into character data.
             */
            this.modifiedCharacterRanges = calculateCharacterRangesFromLineChanges(
                lineChangeData,
                this.fileContents
            );
            return {err: null};
        } catch (err) {
            return {err: assertInstanceOf(err, Error)};
        }
    }
}
