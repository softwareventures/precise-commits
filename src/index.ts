import {isAbsolute, join, relative} from "path";

import type {Observable} from "rxjs";
import {concatAll, count, endWith, filter, from, map, mergeMap, mergeWith, of, scan} from "rxjs";
import {
    getModifiedFilenames,
    index,
    resolveNearestGitDirectoryParent,
    workingTree
} from "./git-utils";
import {generateFilesWhitelistPredicate, noLineChangeDataError} from "./utils";
import {preciseFormatterPrettier} from "./precise-formatters/prettier";
import {ModifiedFile} from "./modified-file";

export interface AdditionalOptions {
    checkOnly: boolean;
    filesWhitelist: string[] | null;
    base: string | null;
    head: string | null;
}

export type State = Initializing | Running | Finished;

export interface Initializing {
    readonly state: "Initializing";
}

export interface Running {
    readonly state: "Running";
    readonly gitSearchComplete: boolean;
    readonly files: readonly FileState[];
}

export interface FileState {
    readonly filename: string;
    readonly status: FileStatus;
}

export type FileStatus = "Processing" | "NotUpdated" | "Updated" | "InvalidFormatting";

export interface Finished {
    readonly state: "Finished";
    readonly fileCount: number;
}

/**
 * LIBRARY
 */
export function main(
    workingDirectory: string,
    additionalOptions: AdditionalOptions
): Observable<State> {
    // Merge user-given and default options.
    return of({
        ...{
            filesWhitelist: null,
            base: null,
            head: null,
            checkOnly: false,
            formatter: "prettier"
        },
        ...additionalOptions
    }).pipe(
        mergeMap(async options => {
            // Note: Will be exposed as an option if/when new formatters are added.
            if (options.formatter !== "prettier") {
                throw new Error(`The only supported value for "formatter" option is "prettier"`);
            }

            const selectedFormatter = preciseFormatterPrettier;

            // Resolve the relevant .git directory's parent directory up front, as we will need this when
            // executing various `git` commands.
            const gitDirectoryParent = await resolveNearestGitDirectoryParent(workingDirectory);

            // Find files that we should process. A file is relevant if:
            //  * The file extension is supported by the given formatter.
            //  * The file is included in the optional `filesWhitelist` array, or no whitelist is specified.
            //  * The file is not ignored as a result of any supported "ignore" mechanism of the formatter.
            const relevantFiles = from(
                getModifiedFilenames(gitDirectoryParent, options.base, options.head)
            ).pipe(
                mergeMap(array => array),
                map(path => join(gitDirectoryParent, path)),
                map(path => relative(workingDirectory, path)),
                filter(path => !isAbsolute(path)),
                filter(selectedFormatter.hasSupportedFileExtension),
                filter(generateFilesWhitelistPredicate(options.filesWhitelist)),
                filter(selectedFormatter.generateIgnoreFilePredicate(workingDirectory))
            );

            const newFileEvents = relevantFiles.pipe(
                map((filename, fileIndex) => ({event: "NewFile", fileIndex, filename} as const))
            );

            const gitSearchCompleteEvent = relevantFiles.pipe(
                count(),
                map(fileCount => ({event: "GitSearchComplete", fileCount} as const))
            );

            const fileProcessedEvents = relevantFiles.pipe(
                mergeMap(async (filename, fileIndex) => {
                    const fullPath = join(workingDirectory, filename);

                    // Read the modified file contents and resolve the relevant formatter.
                    const modifiedFile = new ModifiedFile({
                        fullPath,
                        gitDirectoryParent,
                        base: options.base,
                        head: options.head ?? index,
                        selectedFormatter
                    });

                    // To avoid unnecessary issues with 100% valid files producing issues when parts
                    // of them are reformatted in isolation, we first check the whole file to see if
                    // it is already formatted. This could also allow us to skip unnecessary git diff
                    // analysis work.
                    if (modifiedFile.isAlreadyFormatted()) {
                        return {
                            event: "FileProcessed",
                            fileIndex,
                            filename,
                            status: "NotUpdated"
                        } as const;
                    }

                    // Calculate what character ranges have been affected in the modified file.
                    // If any of the analysis threw an error for any reason, it will be returned
                    // from the method so we can handle it here.
                    const {err} = modifiedFile.calculateModifiedCharacterRanges();
                    if (err != null) {
                        if (err.message === noLineChangeDataError) {
                            return {
                                event: "FileProcessed",
                                fileIndex,
                                filename,
                                status: "NotUpdated"
                            } as const;
                        }

                        // Unexpected error
                        throw err;
                    }

                    // "CHECK ONLY MODE"
                    if (options.checkOnly) {
                        return {
                            event: "FileProcessed",
                            fileIndex,
                            filename,
                            status: modifiedFile.hasValidFormattingForCharacterRanges()
                                ? "NotUpdated"
                                : "InvalidFormatting"
                        } as const;
                    }

                    // "FORMAT MODE"
                    modifiedFile.formatCharacterRangesWithinContents();
                    if (!modifiedFile.shouldContentsBeUpdatedOnDisk()) {
                        return {
                            event: "FileProcessed",
                            fileIndex,
                            filename,
                            status: "NotUpdated"
                        } as const;
                    }

                    // If we're updating the index, we also need to update the working tree.
                    if (options.head == null) {
                        const workingTreeFile = new ModifiedFile({
                            fullPath,
                            gitDirectoryParent,
                            base: options.base,
                            head: workingTree,
                            selectedFormatter
                        });
                        const {err} = workingTreeFile.calculateModifiedCharacterRanges();
                        if (err == null) {
                            workingTreeFile.formatCharacterRangesWithinContents();
                            if (workingTreeFile.shouldContentsBeUpdatedOnDisk()) {
                                workingTreeFile.updateFileOnDisk();
                            }
                        } else if (err.message !== noLineChangeDataError) {
                            // Unexpected error
                            throw err;
                        }
                    }

                    // Write the file back to disk and report.
                    modifiedFile.updateFileOnDisk();
                    return {
                        event: "FileProcessed",
                        fileIndex,
                        filename,
                        status: "Updated"
                    } as const;
                })
            );

            return newFileEvents.pipe(mergeWith(gitSearchCompleteEvent, fileProcessedEvents));
        }),
        concatAll(),
        endWith({event: "Finished"} as const),
        scan(
            (state: State, event) => {
                if (event.event === "NewFile") {
                    if (state.state === "Running") {
                        return {
                            ...state,
                            files: [
                                ...state.files,
                                {
                                    filename: event.filename,
                                    status: "Processing"
                                }
                            ]
                        } as const;
                    } else {
                        return {
                            state: "Running",
                            gitSearchComplete: false,
                            files: [{filename: event.filename, status: "Processing"}]
                        } as const;
                    }
                } else if (event.event === "GitSearchComplete") {
                    if (state.state === "Running") {
                        return {...state, gitSearchComplete: true} as const;
                    } else {
                        return {state: "Running", gitSearchComplete: true, files: []} as const;
                    }
                } else if (event.event === "FileProcessed") {
                    if (state.state === "Running") {
                        const files = [...state.files];
                        files[event.fileIndex] = {
                            filename: event.filename,
                            status: event.status
                        };
                        return {...state, files} as const;
                    } else {
                        return {
                            state: "Running",
                            gitSearchComplete: false,
                            files: [
                                {
                                    filename: event.filename,
                                    status: event.status
                                }
                            ]
                        } as const;
                    }
                } else if (event.event === "Finished") {
                    if (state.state === "Running") {
                        return {state: "Finished", fileCount: state.files.length} as const;
                    } else {
                        return {state: "Finished", fileCount: 0} as const;
                    }
                } else {
                    return state;
                }
            },
            {state: "Initializing"}
        )
    );
}
