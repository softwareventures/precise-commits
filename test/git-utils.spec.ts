import {sep} from "path";
import * as tempy from "tempy";
import {mkdirp} from "mkdirp";
import {runCommand} from "../src/utils";
import {
    getDiffForFile,
    resolveGitWorkingTreePath,
    getModifiedFilenames,
    index
} from "../src/git-utils";
import {TestBed, readFixtures} from "./test-utils";

const fixtures = readFixtures();
let testBed: TestBed;

describe("git-utils", () => {
    describe("resolveGitWorkingTreePath()", () => {
        beforeAll(() => {
            testBed = new TestBed();
        });

        afterAll(async () => {
            await testBed.teardown();
        });

        fixtures.forEach(fixture => {
            it(fixture.fixtureName, async () => {
                await testBed.prepareFixtureInTmpDirectory(fixture);
                const tmpFile = testBed.getTmpFileForFixture(fixture);
                /**
                 * The tmpFile should resolve to its own .git directory
                 */
                expect(await resolveGitWorkingTreePath(tmpFile.directoryPath)).toEqual(
                    tmpFile.directoryPath
                );
            });
        });

        it(`should resolve the correct working tree path in a working tree created by git worktree`, async () => {
            await tempy.directory.task(async worktreePath => {
                await runCommand("git", ["worktree", "add", "-B", "worktree", worktreePath]);
                const subdirPath = `${worktreePath}${sep}subdir`;
                await mkdirp(subdirPath);
                expect(await resolveGitWorkingTreePath(subdirPath)).toEqual(worktreePath);
                await runCommand("git", ["worktree", "remove", worktreePath]);
            });
        });
    });

    describe("getDiffForFile()", () => {
        beforeAll(() => {
            testBed = new TestBed();
        });

        afterAll(async () => {
            await testBed.teardown();
        });

        fixtures.forEach(fixture => {
            it(fixture.fixtureName, async () => {
                await testBed.prepareFixtureInTmpDirectory(fixture);
                const tmpFile = testBed.getTmpFileForFixture(fixture);
                const gitDirectoryParent = await resolveGitWorkingTreePath(tmpFile.directoryPath);
                const diff = await getDiffForFile(
                    gitDirectoryParent,
                    tmpFile.path,
                    tmpFile.initialCommitSHA,
                    tmpFile.updatedCommitSHA ?? index
                );
                expect(diff).toMatchSnapshot();
            });
        });
    });

    describe("getModifiedFilenames()", () => {
        beforeAll(() => {
            testBed = new TestBed();
        });

        afterAll(async () => {
            await testBed.teardown();
        });

        fixtures.forEach(fixture => {
            it(fixture.fixtureName, async () => {
                await testBed.prepareFixtureInTmpDirectory(fixture);
                const tmpFile = testBed.getTmpFileForFixture(fixture);
                const gitDirectoryParent = await resolveGitWorkingTreePath(tmpFile.directoryPath);
                const fileNames = await getModifiedFilenames(
                    gitDirectoryParent,
                    tmpFile.initialCommitSHA,
                    tmpFile.updatedCommitSHA
                );
                expect(fileNames).toEqual([`${tmpFile.filename}`]);
            });
        });
    });
});
