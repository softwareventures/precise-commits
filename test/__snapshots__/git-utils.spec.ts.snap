// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git-utils getDiffForFile() brand-new-json-file 1`] = `
"diff --git a/brand-new-json-file.json b/brand-new-json-file.json
new file mode 100644
index 0000000..f663bcc
--- /dev/null
+++ b/brand-new-json-file.json
@@ -0,0 +1,3 @@
+{
+    "foo": true
+}
\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() brand-new-jsx-file 1`] = `
"diff --git a/brand-new-jsx-file.jsx b/brand-new-jsx-file.jsx
new file mode 100644
index 0000000..ae72777
--- /dev/null
+++ b/brand-new-jsx-file.jsx
@@ -0,0 +1,21 @@
+function HelloWorld({greeting = "hello", greeted = '"World"', silent = false, onMouseOver,}) {
+
+    if(!greeting){return null};
+  
+       // TODO: Don't use random in render
+    let num = Math.floor (Math.random() * 1E+7).toString().replace(/\\.\\d+/ig, "")
+  
+    return <div className='HelloWorld' title={\`You are visitor number \${ num }\`} onMouseOver={onMouseOver}>
+  
+      <strong>{ greeting.slice( 0, 1 ).toUpperCase() + greeting.slice(1).toLowerCase() }</strong>
+      {greeting.endsWith(",") ? " " : <span style={{color: '\\grey'}}>", "</span> }
+      <em>
+      { greeted }
+      </em>
+      { (silent)
+        ? "."
+        : "!"}
+  
+      </div>;
+  
+  }
\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() broken 1`] = `
"diff --git a/broken.ts b/broken.ts
index c3ce682..fb3933a 100644
--- a/broken.ts
+++ b/broken.ts
@@ -3 +3 @@ export class ProductInventoryDataProvider implements DataSource<ProductInventory
-        T extends {customerId: number; categoryOverview?: CategorySummaryDto}
+        T extends {customerId: number; categoryOverview?: CategorySummaryDto | undefined}"
`;

exports[`git-utils getDiffForFile() changed-line-in-closure 1`] = `
"diff --git a/changed-line-in-closure.ts b/changed-line-in-closure.ts
index 6ad59aa..c61b8fb 100644
--- a/changed-line-in-closure.ts
+++ b/changed-line-in-closure.ts
@@ -3,0 +4,5 @@ import {waitFor} from 'wait-for'
+const failOnError = error => {
+    console.log("got error:", error);
+    fail(error);
+}
+
@@ -7,4 +12,5 @@ describe("salad sausage", () => {
-        repo.findAbc(
-            111,
-            Egg.fromText("nettle")
-        )
+        const subscription = repo.findAbc(
+                111,
+                Egg.fromText("nettle")
+        );
+        subscription
@@ -17,3 +23 @@ describe("salad sausage", () => {
-                error => {
-                    console.log("got error:", error);
-                },
+                failOnError,
@@ -22 +26 @@ describe("salad sausage", () => {
-                    console.log("finished");
+                    console.log("finished"); // TODO: winkle ankle cricket
@@ -24,0 +29,3 @@ describe("salad sausage", () => {
+        subscription.getEffects().subscribe(
+                effect => commandEffectsReceived.push(effect)
+        )
@@ -27,0 +35 @@ describe("salad sausage", () => {
+            expect(commandEffectsReceived).toBeEmpty();"
`;

exports[`git-utils getDiffForFile() empty-javascript-file 1`] = `
"diff --git a/empty-javascript-file.js b/empty-javascript-file.js
index e69de29..3772dd0 100644
--- a/empty-javascript-file.js
+++ b/empty-javascript-file.js
@@ -0,0 +1,2 @@
+var greeting = 'hello'
+        var audience = 'world'
\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() formatting-would-affect-total-line-count 1`] = `
"diff --git a/formatting-would-affect-total-line-count.js b/formatting-would-affect-total-line-count.js
index 4e74cec..5eda403 100644
--- a/formatting-would-affect-total-line-count.js
+++ b/formatting-would-affect-total-line-count.js
@@ -5 +5,2 @@ function myfunc(
-    b
+    b,
+    c
@@ -8,0 +10,2 @@ function myfunc(
+
+var newStatement = true"
`;

exports[`git-utils getDiffForFile() hunk-marker-in-code 1`] = `
"diff --git a/hunk-marker-in-code.ts b/hunk-marker-in-code.ts
index 1f462db..917655e 100644
--- a/hunk-marker-in-code.ts
+++ b/hunk-marker-in-code.ts
@@ -2,2 +2 @@ export function extractLineChangeData(diffData: string) {
-    const lineChanges = diffData.match(/@@.*@@/g);
-    return lineChanges;
+    return diffData.match(/@@.*@@/gu );"
`;

exports[`git-utils getDiffForFile() multiple-updated-lines-javascript-file 1`] = `
"diff --git a/multiple-updated-lines-javascript-file.js b/multiple-updated-lines-javascript-file.js
index 50d7ab7..2b98ecd 100644
--- a/multiple-updated-lines-javascript-file.js
+++ b/multiple-updated-lines-javascript-file.js
@@ -2 +2 @@ var     untouchedUglyLines1 = ''
-function TouchedFunctionSignature({greeting = "Hello", greeted = '"World"', silent = false,}) {
+function TouchedFunctionSignature({greeting = "Hello", greeted = '"Mars"', silent = false,}) {
@@ -5,0 +6 @@ function TouchedFunctionSignature({greeting = "Hello", greeted = '"World"', sile
+      const newlyAddedUglyLine = true
\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() single-updated-line-typescript 1`] = `
"diff --git a/single-updated-line-typescript.ts b/single-updated-line-typescript.ts
index 30011dc..1f5fc71 100644
--- a/single-updated-line-typescript.ts
+++ b/single-updated-line-typescript.ts
@@ -2 +2 @@ const a = true
-const b = false
\\ No newline at end of file
+const b = true
\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() updated-class-body 1`] = `
"diff --git a/updated-class-body.js b/updated-class-body.js
index 003605a..9eafeaf 100644
--- a/updated-class-body.js
+++ b/updated-class-body.js
@@ -2,2 +2,2 @@ class Hatstand {
-    constructor () {
-        this.hats = [];
+    constructor (hats) {
+        this.hats = hats ?? [ ];"
`;
